File: backend/models/adjustment.py
"""Ajustement d'un dossier ACTIF (3.63.0) : ajouter des ressources, en retirer et/ou changer le service d'une personne, sur son
dossier existant, sans repartir d'un nouveau dossier.

Choix de conception (voir docs/REPRISE_MAJ.md et la memoire du chantier) :
- le dossier reste `active` et verrouille : un ajustement n'est PAS une restitution (`derive_restitution_workflow_status` calcule
  `returned` / `partial_return`, pense pour une fin de mission) ; son propre etat est porte par l'evenement
  (`derive_adjustment_status`) dans `payload["ajustements"]`, historique horodate de tous les gestes ;
- chaque geste a SA signature : presentiel (geste), a distance (differee, recueillie ensuite), ou impossible (un responsable
  signe a la place : nom + qualite tracés séparément du bénéficiaire) ;
- les retraits reprennent le mecanisme existant (`restitution.items[cle]`) : le parc et le stock se resynchronisent seuls via
  `persist_form`, comme pour une restitution partielle ;
- fonctions pures (aucun acces base) : la route les appelle puis enregistre en UNE fois avec `persist_form`.
"""
from models.workflow import collect_resource_validation_errors, extract_items
from utils import AppError, generate_id, utc_now

SIGNATURE_MODES = ("presentiel", "distance", "impossible")
RETURN_STATES = ("conforme", "degrade", "autre")
MAX_SIGNATURE_BYTES = 1_048_576  # meme plafond que les liens de signature
MAX_TEXT = 500

# Etats d'un evenement d'ajustement (distincts des statuts de dossier)
STATUS_SIGNED = "signed"
STATUS_PENDING = "pending_signature"
STATUS_SIGNED_BY_SUBSTITUTE = "signed_by_substitute"


def derive_adjustment_status(signature):
    """Etat d'un ajustement d'apres sa signature (jamais d'apres le statut du dossier)."""
    status = (signature or {}).get("status")
    if status == "signed":
        return STATUS_SIGNED
    if status == "substitute":
        return STATUS_SIGNED_BY_SUBSTITUTE
    return STATUS_PENDING


def held_resource_keys(payload):
    """Cles des ressources actuellement detenues (attribuees et pas encore rendues)."""
    return {item["item_key"] for item in extract_items(payload) if item["assigned"] and not item["returned"]}


def _text(value, limit=MAX_TEXT):
    return str(value or "").strip()[:limit]


def _resource_key(resource):
    return str(resource.get("code") or resource.get("id") or "").strip()


def _build_signature(raw, now):
    """Signature d'un geste. `raw` = {mode, signatureDataUrl, reason, substitute: {name, quality}}."""
    raw = raw if isinstance(raw, dict) else {}
    mode = raw.get("mode")
    if mode not in SIGNATURE_MODES:
        raise AppError("signature_mode_required", "Choisissez le mode de signature (présentiel, à distance ou impossible).", 400)
    if mode == "presentiel":
        data_url = raw.get("signatureDataUrl") or ""
        if not str(data_url).startswith("data:image/"):
            raise AppError("signature_required", "La signature est obligatoire en présentiel.", 400)
        if len(data_url) > MAX_SIGNATURE_BYTES:
            raise AppError("signature_too_large", "La signature est trop volumineuse.", 413)
        return {"mode": mode, "status": "signed", "signedAt": now, "signatureDataUrl": data_url}
    if mode == "distance":
        return {"mode": mode, "status": "pending"}
    reason = _text(raw.get("reason"))
    substitute = raw.get("substitute") if isinstance(raw.get("substitute"), dict) else {}
    name, quality = _text(substitute.get("name"), 120), _text(substitute.get("quality"), 120)
    if not reason:
        raise AppError("signature_reason_required", "Précisez pourquoi la signature est impossible.", 400)
    if not name:
        raise AppError("substitute_required", "Indiquez qui signe à la place (nom et qualité).", 400)
    return {"mode": mode, "status": "substitute", "signedAt": now, "reason": reason,
            "substitute": {"name": name, "quality": quality}}


def _resource_summary(resource):
    return {"key": _resource_key(resource), "label": _text(resource.get("label"), 200) or _resource_key(resource)}


def apply_adjustment(payload, body, actor, now=None):
    """Applique un ajustement au payload d'un dossier actif. Renvoie (payload, evenement). Leve AppError si invalide."""
    now = now or utc_now()
    body = body if isinstance(body, dict) else {}
    if (payload.get("workflow") or {}).get("status") != "active":
        raise AppError("not_adjustable", "Seul un dossier actif peut être ajusté.", 409)

    held = held_resource_keys(payload)
    additions, withdrawals = [], []
    resources = payload.setdefault("resources", {})
    if not isinstance(resources.get("additional"), list):
        resources["additional"] = []
    seen = set()

    for raw in body.get("ajouts") or []:
        if not isinstance(raw, dict) or not _resource_key(raw):
            raise AppError("invalid_addition", "Chaque ressource ajoutée doit avoir un code.", 400)
        key = _resource_key(raw)
        if key in held or key in seen:
            raise AppError("already_held", f"La ressource « {_text(raw.get('label')) or key} » est déjà détenue.", 409)
        seen.add(key)
        entry = {**raw, "selected": True, "category": raw.get("category") or "materiel"}
        additions.append(entry)

    for raw in body.get("retraits") or []:
        key = _text(raw.get("key") if isinstance(raw, dict) else "", 200)
        if not key or key not in held:
            raise AppError("not_held", "Cette ressource n'est pas détenue : impossible de la retirer.", 409)
        if key in seen:
            raise AppError("duplicate_gesture", "Une ressource ne peut pas être ajoutée et retirée en même temps.", 400)
        seen.add(key)
        state = raw.get("state") or "conforme"
        if state not in RETURN_STATES:
            raise AppError("invalid_state", "État de reprise inconnu.", 400)
        withdrawals.append({"key": key, "state": state, "notes": _text(raw.get("notes"))})

    beneficiaire = payload.setdefault("beneficiaire", {})
    new_service = _text(body.get("service"), 200)
    old_service = beneficiaire.get("service") or ""
    service_change = {"from": old_service, "to": new_service} if new_service and new_service != old_service else None

    if not additions and not withdrawals and not service_change:
        raise AppError("empty_adjustment", "Rien à ajuster : ajoutez ou retirez une ressource, ou changez le service.", 400)

    signature = _build_signature(body.get("signature"), now)
    event_id = generate_id("ajust")

    # Le libelle d'un retrait est celui de la ressource detenue (pas une valeur envoyee par le client).
    labels = {item["item_key"]: item["label"] for item in extract_items(payload)}
    restitution = payload.setdefault("restitution", {})
    items = restitution.get("items")
    if not isinstance(items, dict):
        items = restitution["items"] = {}
    for withdrawal in withdrawals:
        items[withdrawal["key"]] = {"state": withdrawal["state"], "notes": withdrawal["notes"], "returnedAt": now, "adjustmentId": event_id}
    for entry in additions:
        entry["adjustmentId"] = event_id
        entry["assignedAt"] = now
        resources["additional"].append(entry)
    if service_change:
        beneficiaire["service"] = new_service

    errors = collect_resource_validation_errors(payload)
    if errors:
        raise AppError("resource_incomplete", "Une ressource ajoutée n'est pas complètement renseignée : " + " ; ".join(map(str, errors[:3])), 400)

    event = {
        "id": event_id,
        "at": now,
        "by": actor,
        "ajouts": [_resource_summary(entry) for entry in additions],
        "retraits": [{"key": w["key"], "label": labels.get(w["key"], w["key"]), "state": w["state"], "notes": w["notes"]} for w in withdrawals],
        "service": service_change,
        "signature": signature,
        "status": derive_adjustment_status(signature),
    }
    payload.setdefault("ajustements", []).append(event)
    return payload, event


def complete_adjustment_signature(payload, event_id, raw_signature, now=None):
    """Recueille la signature d'un ajustement resté « en attente » (mode à distance). Renvoie (payload, evenement)."""
    now = now or utc_now()
    event = next((e for e in payload.get("ajustements") or [] if e.get("id") == event_id), None)
    if not event:
        raise AppError("not_found", "Ajustement introuvable.", 404)
    if event.get("status") != STATUS_PENDING:
        raise AppError("already_signed", "Cet ajustement est déjà signé.", 409)
    raw = dict(raw_signature) if isinstance(raw_signature, dict) else {}
    if raw.get("mode") == "distance":
        raise AppError("signature_mode_required", "Recueillez la signature (présentiel) ou indiquez qu'elle est impossible.", 400)
    event["signature"] = {**_build_signature(raw, now), "requestedMode": "distance"}
    event["status"] = derive_adjustment_status(event["signature"])
    return payload, event


def public_event(event):
    """Evenement sans l'image de la signature (jamais renvoyee par les routes d'ajustement)."""
    clean = dict(event)
    signature = dict(clean.get("signature") or {})
    signature.pop("signatureDataUrl", None)
    clean["signature"] = signature
    return clean

File: frontend/js/adjustment.js
// Ajustement d'un dossier actif (3.65.0) : ajouter / retirer des ressources et/ou changer le service d'une personne sur son
// dossier existant, avec une signature par geste. Chargé après storage.js (listes et fiche). Le serveur fait foi : toutes les
// règles (dossier actif, ressource détenue, ressource complète, verrou optimiste) sont revérifiées par
// PATCH /api/forms/<id>/ajustement ; ce fichier ne fait que les proposer et afficher les refus.

// Les modes de signature et les états de reprise sont décrits ici, une seule fois, puis générés.
const ADJUSTMENT_SIGNATURE_MODES = [
  { value: "presentiel", label: "En présentiel", hint: "La personne signe sur cet écran." },
  { value: "distance", label: "À distance", hint: "L'ajustement est enregistré « en attente » ; la signature est recueillie ensuite (Ajuster > Signer l'ajustement)." },
  { value: "impossible", label: "Signature impossible", hint: "Un responsable signe à la place : indiquez le motif, son nom et sa qualité." }
];
const ADJUSTMENT_RETURN_STATES = [["conforme", "Conforme"], ["degrade", "Dégradé"], ["autre", "Autre"]];
const ADJUSTMENT_CONDITIONS = [["", "Non précisé"], ["neuf", "Neuf"], ["bon_etat", "Bon état"], ["etat_usage", "État d'usage"], ["degrade", "Dégradé"]];
const ADJUSTMENT_STATUS_LABELS = { signed: "Signé", pending_signature: "Signature en attente", signed_by_substitute: "Signé par un responsable" };
const ADJUSTMENT_ERRORS = {
  form_conflict: "Le dossier a été modifié depuis son ouverture. Fermez cette fenêtre, rouvrez l'ajustement et recommencez.",
  forbidden: "Votre profil ne permet pas d'ajuster un dossier.",
  masked_scope_read_only: "Votre profil (données masquées) ne permet pas d'ajuster un dossier."
};

function adjEsc(value) {
  return String(value ?? "").replace(/[&<>"']/g, (char) => ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[char]));
}

function adjOptions(pairs, selected = "") {
  return pairs.map(([value, label]) => `<option value="${adjEsc(value)}"${value === selected ? " selected" : ""}>${adjEsc(label)}</option>`).join("");
}

// ── Signature manuscrite minimale (canvas), indépendante de celles de la fiche et de la restitution ──
function createAdjustmentSignaturePad(canvas) {
  const context = canvas.getContext("2d");
  let drawing = false;
  let dirty = false;
  const resize = () => {
    const ratio = window.devicePixelRatio || 1;
    const width = Math.round((canvas.clientWidth || 400) * ratio);
    const height = Math.round((canvas.clientHeight || 160) * ratio);
    // Redimensionner efface le dessin : on ne le fait que si la taille change, et l'indicateur « dessiné » suit.
    if (canvas.width === width && canvas.height === height) return;
    canvas.width = width;
    canvas.height = height;
    context.setTransform(ratio, 0, 0, ratio, 0, 0);
    context.lineWidth = 2;
    context.lineCap = "round";
    context.strokeStyle = "#111";
    dirty = false;
  };
  const point = (event) => {
    const rect = canvas.getBoundingClientRect();
    return { x: event.clientX - rect.left, y: event.clientY - rect.top };
  };
  canvas.addEventListener("pointerdown", (event) => {
    drawing = true;
    canvas.setPointerCapture?.(event.pointerId);
    const { x, y } = point(event);
    context.beginPath();
    context.moveTo(x, y);
    event.preventDefault();
  });
  canvas.addEventListener("pointermove", (event) => {
    if (!drawing) return;
    const { x, y } = point(event);
    context.lineTo(x, y);
    context.stroke();
    dirty = true;
  });
  ["pointerup", "pointercancel", "pointerleave"].forEach((name) => canvas.addEventListener(name, () => { drawing = false; }));
  resize();
  return {
    isEmpty: () => !dirty,
    clear: () => { context.clearRect(0, 0, canvas.width, canvas.height); dirty = false; },
    toDataUrl: () => (dirty ? canvas.toDataURL("image/png") : ""),
    resize
  };
}

// ── Droits et lecture de l'historique ──
function getAdjustments(draft) {
  const events = draft?.data?.ajustements ?? draft?.ajustements;
  return Array.isArray(events) ? events : [];
}

function getPendingAdjustment(draft) {
  return getAdjustments(draft).find((event) => event.status === "pending_signature") || null;
}

function describeAdjustment(event) {
  const parts = [];
  (event.ajouts || []).forEach((item) => parts.push(`+ ${item.label}`));
  (event.retraits || []).forEach((item) => parts.push(`− ${item.label}`));
  if (event.service) parts.push(`Service : ${event.service.from || "—"} → ${event.service.to}`);
  return parts;
}

function renderAdjustmentHistory(container, events) {
  if (!container) return;
  const list = Array.isArray(events) ? events : [];
  container.classList.toggle("d-none", !list.length);
  if (!list.length) {
    container.innerHTML = "";
    return;
  }
  container.innerHTML = `
    <p class="panel-eyebrow mb-1">Historique des ajustements</p>
    <ul class="list-group">
      ${list.slice().reverse().map((event) => {
        const when = event.at ? new Date(event.at).toLocaleString("fr-FR", { dateStyle: "short", timeStyle: "short" }) : "";
        const substitute = event.signature?.substitute ? ` — ${adjEsc(event.signature.substitute.name)} (${adjEsc(event.signature.substitute.quality || "responsable")})` : "";
        return `<li class="list-group-item">
          <div class="d-flex justify-content-between gap-2"><strong>${adjEsc(when)}</strong><span class="badge text-bg-secondary">${adjEsc(ADJUSTMENT_STATUS_LABELS[event.status] || event.status)}${substitute}</span></div>
          <div class="small">${describeAdjustment(event).map(adjEsc).join("<br>")}</div>
          <div class="small text-muted">par ${adjEsc(event.by || "—")}</div>
        </li>`;
      }).join("")}
    </ul>`;
}

// ── Champs d'une ressource ajoutée (mêmes types que la fiche : texte, nombre, date, liste, choix, case à cocher, liste de valeurs) ──
function buildAdjustmentFieldInput(resource, field) {
  const id = `adj_${resource.id}_${field.key}`;
  const label = `<label class="form-label" for="${adjEsc(id)}">${adjEsc(field.label)}${field.required ? " *" : ""}</label>`;
  const attrs = `id="${adjEsc(id)}" data-adj-field="${adjEsc(field.key)}" data-adj-type="${adjEsc(field.type || "text")}"${field.required ? " required" : ""}`;
  if (field.type === "textarea" || field.type === "list") {
    const help = field.type === "list" ? '<div class="form-text">Une valeur par ligne.</div>' : "";
    return `<div class="mb-2">${label}<textarea class="form-control" rows="2" ${attrs}></textarea>${help}</div>`;
  }
  if (field.type === "select") {
    const options = (Array.isArray(field.options) ? field.options : []).map((option) => [option, option]);
    return `<div class="mb-2">${label}<select class="form-select" ${attrs}><option value="">Sélectionner</option>${adjOptions(options)}</select></div>`;
  }
  if (field.type === "checkbox") {
    return `<div class="mb-2"><label class="form-check"><input class="form-check-input" type="checkbox" ${attrs}><span class="form-check-label">${adjEsc(field.label)}</span></label></div>`;
  }
  const inputType = ["number", "date"].includes(field.type) ? field.type : "text";
  return `<div class="mb-2">${label}<input class="form-control" type="${inputType}" ${attrs}></div>`;
}

function buildAdjustmentAdditionBlock(resource) {
  const visible = (Array.isArray(resource.field_schema) ? resource.field_schema : []).filter((field) => !field.hidden);
  const fields = visible.length
    ? visible.map((field) => buildAdjustmentFieldInput(resource, field)).join("")
    : `<div class="mb-2"><label class="form-label" for="adj_${adjEsc(resource.id)}_details">Précision / détails</label><input class="form-control" id="adj_${adjEsc(resource.id)}_details" data-adj-details="true" placeholder="Numéro de série, taille, couleur…"></div>`;
  const condition = resource.has_assignment_condition
    ? `<div class="mb-2"><label class="form-label">État à la remise</label><select class="form-select" data-adj-condition="true">${adjOptions(ADJUSTMENT_CONDITIONS)}</select></div>` : "";
  return `<div class="border rounded p-3 mb-2" data-adj-addition="${adjEsc(resource.id)}">
    <div class="d-flex justify-content-between align-items-start mb-2"><strong>+ ${adjEsc(resource.label)}</strong>
      <button type="button" class="btn btn-sm btn-outline-secondary" data-adj-remove-addition="${adjEsc(resource.id)}">Annuler l'ajout</button></div>
    ${fields}${condition}</div>`;
}

function readAdjustmentAddition(block, resource) {
  const fields = {};
  block.querySelectorAll("[data-adj-field]").forEach((input) => {
    const type = input.dataset.adjType;
    let value = input.type === "checkbox" ? (input.checked ? "Oui" : "") : input.value.trim();
    if (type === "list") value = value.split("\n").map((line) => line.trim()).filter(Boolean);
    if (Array.isArray(value) ? value.length : value) fields[input.dataset.adjField] = value;
  });
  const missing = (Array.isArray(resource.field_schema) ? resource.field_schema : [])
    .filter((field) => !field.hidden && field.required && !String(fields[field.key] ?? "").trim());
  if (missing.length) {
    throw new Error(`« ${resource.label} » : renseignez ${missing.map((field) => `« ${field.label} »`).join(", ")}.`);
  }
  const details = block.querySelector("[data-adj-details]")?.value.trim() || "";
  if (!(Array.isArray(resource.field_schema) && resource.field_schema.length) && !details) {
    throw new Error(`« ${resource.label} » : précisez le détail de la ressource.`);
  }
  return {
    id: resource.id, code: resource.code, label: resource.label, description: resource.description || "", category: resource.category,
    issuerService: resource.issuer_service, requiresReturn: Boolean(resource.requires_return),
    hasAssignmentDate: Boolean(resource.has_assignment_date), hasAssignmentCondition: Boolean(resource.has_assignment_condition),
    hasAssignmentNotes: Boolean(resource.has_assignment_notes), fieldSchema: resource.field_schema || [], fields, details,
    conditionAttribution: block.querySelector("[data-adj-condition]")?.value || ""
  };
}

// ── Fenêtre d'ajustement ──
function ensureAdjustmentModal() {
  let modal = document.getElementById("adjustmentModal");
  if (!modal) {
    modal = document.createElement("div");
    modal.className = "password-generator-modal d-none";
    modal.id = "adjustmentModal";
    modal.setAttribute("aria-hidden", "true");
    document.body.appendChild(modal);
  }
  return modal;
}

function closeAdjustmentModal() {
  const modal = document.getElementById("adjustmentModal");
  if (!modal) return;
  modal.classList.add("d-none");
  modal.setAttribute("aria-hidden", "true");
  modal.innerHTML = "";
}

function buildAdjustmentSignatureSection(prefix) {
  return `<fieldset class="mb-3"><legend class="form-label fs-6">Signature</legend>
    ${ADJUSTMENT_SIGNATURE_MODES.map((mode, index) => `
      <div class="form-check"><input class="form-check-input" type="radio" name="${prefix}_mode" id="${prefix}_mode_${mode.value}" value="${mode.value}"${index === 0 ? " checked" : ""}>
      <label class="form-check-label" for="${prefix}_mode_${mode.value}">${adjEsc(mode.label)} <span class="text-muted small">— ${adjEsc(mode.hint)}</span></label></div>`).join("")}
    <div class="mt-2" id="${prefix}_presentiel"><canvas id="${prefix}_canvas" class="signature-box" style="width:100%;height:160px"></canvas>
      <button type="button" class="btn btn-sm btn-outline-secondary mt-1" id="${prefix}_clear">Effacer la signature</button></div>
    <div class="mt-2 d-none" id="${prefix}_impossible">
      <label class="form-label" for="${prefix}_reason">Motif *</label><input class="form-control mb-2" id="${prefix}_reason" placeholder="Pourquoi la signature est impossible">
      <div class="row g-2"><div class="col-sm-6"><label class="form-label" for="${prefix}_sub_name">Nom du responsable *</label><input class="form-control" id="${prefix}_sub_name"></div>
      <div class="col-sm-6"><label class="form-label" for="${prefix}_sub_quality">Qualité</label><input class="form-control" id="${prefix}_sub_quality" placeholder="Responsable de service"></div></div>
    </div></fieldset>`;
}

function bindAdjustmentSignatureSection(modal, prefix, allowedModes) {
  const pad = createAdjustmentSignaturePad(modal.querySelector(`#${prefix}_canvas`));
  const refresh = () => {
    const mode = modal.querySelector(`input[name="${prefix}_mode"]:checked`)?.value;
    modal.querySelector(`#${prefix}_presentiel`).classList.toggle("d-none", mode !== "presentiel");
    modal.querySelector(`#${prefix}_impossible`).classList.toggle("d-none", mode !== "impossible");
    if (mode === "presentiel") pad.resize();
  };
  modal.querySelectorAll(`input[name="${prefix}_mode"]`).forEach((radio) => {
    if (allowedModes && !allowedModes.includes(radio.value)) radio.closest(".form-check").classList.add("d-none");
    radio.addEventListener("change", refresh);
  });
  modal.querySelector(`#${prefix}_clear`).addEventListener("click", () => pad.clear());
  refresh();
  return () => {
    const mode = modal.querySelector(`input[name="${prefix}_mode"]:checked`)?.value;
    if (mode === "presentiel") {
      if (pad.isEmpty()) throw new Error("La signature est obligatoire en présentiel.");
      return { mode, signatureDataUrl: pad.toDataUrl() };
    }
    if (mode === "impossible") {
      const reason = modal.querySelector(`#${prefix}_reason`).value.trim();
      const name = modal.querySelector(`#${prefix}_sub_name`).value.trim();
      if (!reason) throw new Error("Précisez pourquoi la signature est impossible.");
      if (!name) throw new Error("Indiquez qui signe à la place (nom du responsable).");
      return { mode, reason, substitute: { name, quality: modal.querySelector(`#${prefix}_sub_quality`).value.trim() } };
    }
    return { mode };
  };
}

function adjustmentErrorMessage(error) {
  return ADJUSTMENT_ERRORS[error?.message] || error?.serverMessage || error?.message || "L'ajustement n'a pas pu être enregistré.";
}

async function adjustmentRequest(url, method, body) {
  const response = await fetch(url, {
    method, credentials: "same-origin", cache: "no-store",
    headers: { "Content-Type": "application/json", "X-CSRF-Token": typeof getCsrfToken === "function" ? await getCsrfToken() : "" },
    body: JSON.stringify(body)
  });
  const payload = await response.json().catch(() => ({}));
  if (!response.ok) {
    const error = new Error(payload.error || `HTTP ${response.status}`);
    error.serverMessage = payload.message;
    throw error;
  }
  return payload;
}

function afterAdjustment(message) {
  showToast(message, "success");
  closeAdjustmentModal();
  if (document.getElementById("draftList") && typeof renderDraftList === "function") {
    void renderDraftList();
  } else {
    window.location.reload();
  }
}

async function openAdjustment(id) {
  let dossier;
  let catalog = [];
  let services = [];
  try {
    dossier = await requestJson(`/api/forms/${encodeURIComponent(id)}`);
    [catalog, services] = await Promise.all([
      requestJson("/api/reference/resources").catch(() => []),
      requestJson("/api/reference/services").catch(() => [])
    ]);
  } catch (error) {
    showToast("Impossible de charger le dossier.", "error");
    return;
  }
  if ((dossier.summary?.status || dossier.data?.workflow?.status) !== "active") {
    showToast("Seul un dossier actif peut être ajusté.", "warning");
    return;
  }
  const data = dossier.data || {};
  const held = (dossier.items || []).filter((item) => item.assigned && !item.returned);
  const heldKeys = new Set(held.map((item) => item.itemKey));
  const addable = (Array.isArray(catalog) ? catalog : []).filter((resource) => !heldKeys.has(resource.code));
  const currentService = data.beneficiaire?.service || "";
  const serviceNames = [...new Set([currentService, ...(Array.isArray(services) ? services : []).map((service) => service.name || service.label || service)].filter(Boolean))];
  const baseSavedAt = data.meta?.savedAt || dossier.summary?.updatedAt || "";
  const title = `${data.beneficiaire?.prenom || ""} ${data.beneficiaire?.nom || ""}`.trim() || "Dossier";

  const modal = ensureAdjustmentModal();
  modal.innerHTML = `
    <div class="password-generator-modal__backdrop" data-adj-close="true"></div>
    <div class="password-generator-modal__dialog" role="dialog" aria-modal="true" aria-labelledby="adjustmentTitle">
      <div class="password-generator-modal__header">
        <div><p class="panel-eyebrow">Dossier actif</p><h2 class="section-title" id="adjustmentTitle">Ajuster : ${adjEsc(title)}</h2></div>
        <button class="btn btn-outline-secondary btn-sm" type="button" data-adj-close="true">Fermer</button>
      </div>
      <form class="password-generator-modal__content" id="adjustmentForm" novalidate>
        <fieldset class="mb-3"><legend class="form-label fs-6">Ressources à retirer</legend>
          ${held.length ? held.map((item, index) => `
            <div class="border rounded p-2 mb-2" data-adj-held="${adjEsc(item.itemKey)}">
              <label class="form-check"><input class="form-check-input" type="checkbox" id="adj_held_${index}" data-adj-withdraw="${adjEsc(item.itemKey)}"><span class="form-check-label">${adjEsc(item.label)}</span></label>
              <div class="row g-2 mt-1 d-none" data-adj-withdraw-details>
                <div class="col-sm-4"><select class="form-select form-select-sm" data-adj-state aria-label="État à la reprise">${adjOptions(ADJUSTMENT_RETURN_STATES)}</select></div>
                <div class="col-sm-8"><input class="form-control form-control-sm" data-adj-notes placeholder="Remarque (facultatif)"></div>
              </div>
            </div>`).join("") : '<p class="text-muted small mb-0">Aucune ressource détenue.</p>'}
        </fieldset>
        <fieldset class="mb-3"><legend class="form-label fs-6">Ressource à ajouter</legend>
          <select class="form-select mb-2" id="adjAddResource" aria-label="Ajouter une ressource"><option value="">Ajouter une ressource…</option>
            ${adjOptions(addable.map((resource) => [String(resource.id), `${resource.label}${resource.category ? ` (${resource.category})` : ""}`]))}</select>
          <div id="adjAdditions"></div>
        </fieldset>
        <div class="mb-3"><label class="form-label" for="adjService">Service</label>
          <select class="form-select" id="adjService">${adjOptions(serviceNames.map((name) => [name, name]), currentService)}</select></div>
        ${buildAdjustmentSignatureSection("adjSig")}
        <p class="text-danger small d-none" id="adjustmentError" role="alert"></p>
        <div class="password-generator-modal__actions password-generator-modal__actions--sticky">
          <button class="btn btn-outline-secondary" type="button" data-adj-close="true">Annuler</button>
          <button class="btn btn-primary" type="submit" id="adjustmentSubmit">Enregistrer l'ajustement</button>
        </div>
        <div class="mt-3" id="adjustmentHistory"></div>
      </form>
    </div>`;
  modal.classList.remove("d-none");
  modal.setAttribute("aria-hidden", "false");
  renderAdjustmentHistory(modal.querySelector("#adjustmentHistory"), data.ajustements);

  const readSignature = bindAdjustmentSignatureSection(modal, "adjSig");
  const errorNode = modal.querySelector("#adjustmentError");
  const showError = (message) => { errorNode.textContent = message; errorNode.classList.toggle("d-none", !message); };
  modal.querySelectorAll("[data-adj-close]").forEach((button) => button.addEventListener("click", closeAdjustmentModal));
  modal.querySelectorAll("[data-adj-withdraw]").forEach((box) => box.addEventListener("change", () => {
    box.closest("[data-adj-held]").querySelector("[data-adj-withdraw-details]").classList.toggle("d-none", !box.checked);
  }));
  const additions = modal.querySelector("#adjAdditions");
  const chooser = modal.querySelector("#adjAddResource");
  chooser.addEventListener("change", () => {
    const resource = addable.find((entry) => String(entry.id) === chooser.value);
    chooser.value = "";
    if (!resource || additions.querySelector(`[data-adj-addition="${CSS.escape(String(resource.id))}"]`)) return;
    additions.insertAdjacentHTML("beforeend", buildAdjustmentAdditionBlock(resource));
    const block = additions.lastElementChild;
    block.querySelector("[data-adj-remove-addition]").addEventListener("click", () => block.remove());
  });

  modal.querySelector("#adjustmentForm").addEventListener("submit", async (event) => {
    event.preventDefault();
    showError("");
    const submit = modal.querySelector("#adjustmentSubmit");
    try {
      const withdrawals = [...modal.querySelectorAll("[data-adj-withdraw]:checked")].map((box) => {
        const row = box.closest("[data-adj-held]");
        return { key: box.dataset.adjWithdraw, state: row.querySelector("[data-adj-state]").value, notes: row.querySelector("[data-adj-notes]").value.trim() };
      });
      const newResources = [...additions.querySelectorAll("[data-adj-addition]")].map((block) =>
        readAdjustmentAddition(block, addable.find((entry) => String(entry.id) === block.dataset.adjAddition)));
      const service = modal.querySelector("#adjService").value;
      const serviceChanged = service && service !== currentService;
      if (!withdrawals.length && !newResources.length && !serviceChanged) {
        throw new Error("Rien à ajuster : retirez ou ajoutez une ressource, ou changez le service.");
      }
      const signature = readSignature();
      submit.disabled = true;
      const result = await adjustmentRequest(`/api/forms/${encodeURIComponent(id)}/ajustement`, "PATCH",
        { ajouts: newResources, retraits: withdrawals, service: serviceChanged ? service : "", signature, baseSavedAt });
      afterAdjustment(result.ajustement?.status === "pending_signature"
        ? "Ajustement enregistré : la signature reste à recueillir." : "Ajustement enregistré.");
    } catch (error) {
      submit.disabled = false;
      showError(adjustmentErrorMessage(error));
    }
  });
}

// ── Recueillir la signature d'un ajustement resté « en attente » (mode à distance) ──
async function openAdjustmentSignature(id) {
  let dossier;
  try {
    dossier = await requestJson(`/api/forms/${encodeURIComponent(id)}`);
  } catch (error) {
    showToast("Impossible de charger le dossier.", "error");
    return;
  }
  const pending = getPendingAdjustment(dossier);
  if (!pending) {
    showToast("Aucun ajustement en attente de signature.", "info");
    return;
  }
  const modal = ensureAdjustmentModal();
  modal.innerHTML = `
    <div class="password-generator-modal__backdrop" data-adj-close="true"></div>
    <div class="password-generator-modal__dialog" role="dialog" aria-modal="true" aria-labelledby="adjustmentTitle">
      <div class="password-generator-modal__header">
        <div><p class="panel-eyebrow">Ajustement en attente</p><h2 class="section-title" id="adjustmentTitle">Recueillir la signature</h2></div>
        <button class="btn btn-outline-secondary btn-sm" type="button" data-adj-close="true">Fermer</button>
      </div>
      <form class="password-generator-modal__content" id="adjustmentForm" novalidate>
        <div class="mb-3">${describeAdjustment(pending).map(adjEsc).join("<br>")}</div>
        ${buildAdjustmentSignatureSection("adjSign")}
        <p class="text-danger small d-none" id="adjustmentError" role="alert"></p>
        <div class="password-generator-modal__actions password-generator-modal__actions--sticky">
          <button class="btn btn-outline-secondary" type="button" data-adj-close="true">Annuler</button>
          <button class="btn btn-primary" type="submit" id="adjustmentSubmit">Enregistrer la signature</button>
        </div>
      </form>
    </div>`;
  modal.classList.remove("d-none");
  modal.setAttribute("aria-hidden", "false");
  const readSignature = bindAdjustmentSignatureSection(modal, "adjSign", ["presentiel", "impossible"]);
  modal.querySelector('input[name="adjSign_mode"][value="presentiel"]').checked = true;
  modal.querySelector('input[name="adjSign_mode"][value="presentiel"]').dispatchEvent(new Event("change"));
  const errorNode = modal.querySelector("#adjustmentError");
  modal.querySelectorAll("[data-adj-close]").forEach((button) => button.addEventListener("click", closeAdjustmentModal));
  modal.querySelector("#adjustmentForm").addEventListener("submit", async (event) => {
    event.preventDefault();
    errorNode.classList.add("d-none");
    const submit = modal.querySelector("#adjustmentSubmit");
    try {
      const signature = readSignature();
      submit.disabled = true;
      await adjustmentRequest(`/api/forms/${encodeURIComponent(id)}/ajustement/${encodeURIComponent(pending.id)}/signature`, "POST", signature);
      afterAdjustment("Signature de l'ajustement enregistrée.");
    } catch (error) {
      submit.disabled = false;
      errorNode.textContent = adjustmentErrorMessage(error);
      errorNode.classList.remove("d-none");
    }
  });
}

document.addEventListener("keydown", (event) => {
  if (event.key === "Escape" && document.getElementById("adjustmentModal")?.classList.contains("d-none") === false) closeAdjustmentModal();
});
